tirocks-sys/src/lib.rs

```diff
@@ -13,20 +13,196 @@ mod bindings {
+    rocksdb_Slice {
+        data_: src.as_ptr() as _,
+        size_: src.len(),
     }
 }
 
+/// Convert a rocksdb slice to rust slice.
+///
+/// # Safety
+///
+/// Caller should guarantee the data lives as long as the returned slice.
+// Because the layout of rocksdb_Slice is exact the same as &[u8], this function
+// will be optimized away by compiler.
+#[inline]
+pub unsafe fn s<'a>(src: rocksdb_Slice) -> &'a [u8] {
+    std::slice::from_raw_parts(src.data_ as _, src.size_)
+}
+
+impl rocksdb_Status {
+    #[inline]
+    pub fn with_code(code: rocksdb_Status_Code) -> rocksdb_Status {
+        rocksdb_Status {
+            code_: code,
+            subcode_: rocksdb_Status_SubCode::kNone,
+            sev_: rocksdb_Status_Severity::kNoError,
+            state_: std::ptr::null(),
+        }
+    }
+
+    /// Build a status with given errors.
+    ///
+    /// # Panics
+    ///
+    /// `code` should not be `kOK`.
+    #[inline]
+    pub fn with_error(code: rocksdb_Status_Code, state: impl AsRef<[u8]>) -> rocksdb_Status {
+        assert_ne!(code, rocksdb_Status_Code::kOk);
+        let state = state.as_ref();
+        let state_ = if !state.is_empty() {
+            unsafe { crocksdb_to_cplus_array(r(state)) }
+        } else {
+            std::ptr::null()
+        };
+        rocksdb_Status {
+            code_: code,
+            subcode_: rocksdb_Status_SubCode::kNone,
+            sev_: rocksdb_Status_Severity::kNoError,
+            state_,
+        }
+    }
+
+    #[inline]
+    fn clear_state(&mut self) {
+        if self.state_.is_null() {
+            return;
+        }
+        unsafe {
+            crocksdb_free_cplus_array(self.state_);
+            self.state_ = std::ptr::null();
+        }
+    }
+
+    #[inline]
+    pub fn ok(&self) -> bool {
+        self.code_ == rocksdb_Status_Code::kOk
+    }
+
+    #[inline]
+    pub fn state(&self) -> Option<&[u8]> {
+        if self.state_.is_null() {
+            None
+        } else {
+            unsafe { Some(CStr::from_ptr(self.state_).to_bytes()) }
+        }
+    }
+
+    #[inline]
+    pub fn message(&self) -> Result<Option<&str>, Utf8Error> {
+        if self.state_.is_null() {
+            Ok(None)
+        } else {
+            unsafe { CStr::from_ptr(self.state_).to_str().map(Some) }
+        }
+    }
+
+    #[inline]
+    pub fn code(&self) -> rocksdb_Status_Code {
+        self.code_
+    }
+
+    #[inline]
+    pub fn severity(&self) -> rocksdb_Status_Severity {
+        self.sev_
+    }
+
+    #[inline]
+    pub fn set_severity(&mut self, severity: rocksdb_Status_Severity) {
+        self.sev_ = severity;
+    }
+
+    #[inline]
+    pub fn sub_code(&self) -> rocksdb_Status_SubCode {
+        self.subcode_
+    }
+
+    #[inline]
+    pub fn set_sub_code(&mut self, sub_code: rocksdb_Status_SubCode) {
+        self.subcode_ = sub_code;
+    }
+}
+
+impl Drop for rocksdb_Status {
+    #[inline]
+    fn drop(&mut self) {
+        self.clear_state();
```

Comment:
I presume C++ destructor will be called?

Reply:
`Drop` is called on the Rust side, and Rust know nothing about C++ destructor. `clear_state` invokes an API to free the state, which guarantee the memory is freed by the same allocator.

Reply:
I think it's more complicated than that. E.g. RocksDB could use C++ method to re-allocate the internal state or return a newly created object, Rust could also alter the C++ status in error handler. [This document](https://anssi-fr.github.io/rust-guide/07_ffi.html#memory-and-resource-management) suggests we should stick to allocating at one side.

Reply:
> RocksDB could use C++ method to re-allocate the internal state or return a newly created object

It doesn't matter as long as it's not passed to Rust side. After being passed, it's out of C++ scope, only Rust's logic is involved and internal state will not be changed until being dropped.

> Rust could also alter the C++ status in error handler.

How? Only constructor is provided, so it can only construct a new status instead of modifying it in safe Rust. And in `with_error`, the state is allocated and copied by a C API, which uses the C++ default allocator, just the same as `clear_state`.

Reply:
Oh, I just realized you removed those setters. But for the first case, any status you get through the C interface will be constructed by RocksDB and freed by Rust.

Reply:
> any status you get through the C interface will be constructed by RocksDB and freed by Rust.

Yes. As explained above, `state_` is freed by `clear_state`, which calls `crocksdb_free_cplus_array`, and it calls `delete[]`:

https://github.com/BusyJay/tirocks/blob/c5800e1e4bfaea946e191b308179acb10dc54218/tirocks-sys/crocksdb/c.cc#L6602

 It matches how it's allocated in `Status`:

https://github.com/tikv/rocksdb/blob/f236fe443d06337ca2538414c62f70ba17b8e409/util/status.cc#L56.

 Note `Status` itself doesn't need to be freed as it lives in stack.